### ISSUE TITLE
On mobile devices, full size images are loaded

### ISSUE THREAD

**Issue by grig-23** [woocommerce/woocommerce#28916]
**Prerequisites (mark completed items with an [x]):**
- [x] We have have carried out troubleshooting steps and we believe I have found a bug.
- [x] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

**Describe the bug**
On mobile devices, full size images are loaded

**Expected behavior**
Ideally, optimized images should be loaded as per [this thread](https://a8c.slack.com/archives/C7U3Y3VMY/p1611261041018800).

**Actual behavior**
Full size images are loaded.

**Steps to reproduce the bug (We need to be able to reproduce the bug in order to fix it.)**
Steps to reproduce the bug:
1. With Storefront theme enabled, check the image sizes on a mobile device.

**Isolating the problem (mark completed items with an [x]):**
- [x] We have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] We can reproduce this bug consistently using the steps above.

Reported in 3654598-zen

**Comment by mfs-mindsize** [comment]
@grigorijurasov Ahhhh images and sizes :)

1. Can you specify which images this is effecting?
2. I wasn't able to access the "per this thread". Maybe a screenshot, or a copy+paste?

**Comment by johndcoy** [comment]
@mfs-mindsize 

Product Images loaded anywhere. From the product pages to the home page with product thumbnails. On mobile, Chrome is calling for original product images instead of resized, 416x416 or 324x324 thumbnails.

screenshot: https://d.pr/i/rEJsxL

**Comment by juliaamosova** [comment]
Hi @grigorijurasov,

Thank you for taking the time to report this bug, we really appreciate your help. I can reproduce it on my end. 

I am making this with `high` priority in order for us to fix it soon as it seems to affect performance.

**Comment by pacmanito** [comment]
The issue was given a "priority: high" label, but in 8 months nothing has been done. Can you please elaborate if this is at least on some to-do list?

**Comment by nicholasudell** [comment]
I'm shocked this has been open for almost 3 years now when it is causing extreme slowdowns on mobile and negative ranking in Google Page Speed Insights. Is there any update at all on this work?

Here only four product thumbnails are causing a 1.3s slowdown to home page load speed because Woo Commerce is loading the full sized images on mobile.

![image](https://github.com/woocommerce/woocommerce/assets/12933955/3374bb9e-d86a-4349-a2ff-20bd57ab75d0)

**Comment by mejsler** [comment]
The fact that this issue is still open after 3 years is a mind boggling

**Comment by samueljseay** [comment]
It's tough with the current issue description and comments to track down specific issues we need to solve here. My goal is to clarify where the problems lie with specific reproductions if possible. If you have examples of where my assessment below is incorrect can you comment with reproduction instructions? This will help me track down where we're loading full sized images and should not be. Unfortunately most discussion so far has been at a very high level so it's hard for me to track down and make sure that if we made a change we would actually fix the issue.

In my testing, in most cases we do not load full size images. I'm testing with TT4 theme with some sample products. In most places afaict we serve images with srcset. It pretty much always defaults to the 300x300 image size on mobile screen sizes I tested.

### Summary

🟢 Shop page: uses srcset, loads 300x300 image on modern device size, no complaint from pagespeed
🟢 Products by category: uses srcset, loads 300x300 image on modern device size, no complaint from pagespeed
🟢 Cart block: does not use srcset but loads 300x300 image on modern device size, no complaint from pagespeed
🟡 Single product block, Product Collection block: does use srcset, but according to https://ausi.github.io/respimagelint/ we do not load the correct image size. On my testing of a single product block, page speed says we could save ~33KB (per image) in this instance. (This will give you the "properly size images" warning, but to be clear we're not serving the *full size image*). We certainly should improve this!
🟡 Product Gallery (Beta) Block , same issue, does use srcset but sets the wrong image dimensions when the images are in the preview where they're very small, we should fix this.
🔴 Single product page, we do load the full image, I suspect this is to support how we currently allow users to zoom an image. This could be considered as a separate enhancement to optimize or re-think how we support zoom, I'm sure a more modern implementation would lazy load the full sized image.

Bonus points from pagespeed insights would also be to use more modern image compression formats like WebP.

Does this summarize the current state or have I missed something? Thanks all for your patience 🙇🏻

**Comment by tjcafferkey** [comment]
Great investigation and summary @samueljseay. Considering this is labelled as a high priority issue I'm going to ping a few folks to see if we can come up with some direction based on your findings, or if we need to revisit its priority.

cc @sunyatasattva @danielwrobert @jarekmorawski

**Comment by Aljullu** [comment]
I was taking a look at this as well, adding some findings on top of @samueljseay's awesome early exploration.

> Single product page

The full-size image is downloaded by `jquery.zoom.js`, when it creates the ["zoomed-in" image element](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/client/legacy/js/zoom/jquery.zoom.js#L84-L233). So, in theory, it's not really blocking the initial page render. I thought about some things we could do to improve the current situation, but I feel all of them come with a cost, so it's a difficult decision.

In any case, here there are some ideas:

1. Downloading the full-size image only when the user interacts with the zoom. This has the drawback that it will probably add a delay until the zoomed-in image can be shown, as it won't have been downloaded before.
2. Instead of using the full-size image for the zoom effect, use a smaller size (but still bigger than the thumbnail). This has the drawback that it would change how things work for existing stores, with the risk of breaking things. But I think it might be good to experiment with it if we can make it so it doesn't break backwards-compatibility.
3. We could add `loading="lazy"` to the image. I don't think it would be a big improvement, as in most stores the product image is above-the-fold, so it would render on initial load anyway. But it might be worth trying.

>  Product Collection block

From what I saw, we are setting the `srcset` correctly, the issue is that there aren't enough useful thumbnails for each product image to give much benefit. The issue is specially noticeable for any block that uses "Cropped" product images.

But stepping back for a second: when a new product image is created, these thumbnails are generated:

- WP default image thumbnails: they all keep the original image aspect ratio.
- WC image thumbnails:
  - `woocommerce_single` keeps the aspect ratio.
  - `woocommerce_thumbnail` and `woocommerce_gallery_thumbnail` enforces a specific aspect ratio. By default, a square, but it can be changed in `/wp-admin/customize.php` > WooCommerce > Product Images.

> [!NOTE]
> Sidenote: another thing to keep in mind is that `/wp-admin/customize.php` is not accessible from the UI if the store is using a block theme. This has been a long-time known issue, but maybe it's time to prioritize it? It will require some design/product work to decide where to move the Customizer settings.

When an image is rendered, the `srcset` attribute is filled with the thumbnails that have been generated. So in the case of a Product Collection block with "Full Size" images, any thumbnail that kept the aspect ratio can be used, that's `woocommerce_single` and WP thumbnails, which is quite good. However, if the Product Collection is set to display "Cropped" images, only the `woocommerce_gallery_thumbnail` and `woocommerce_thumbnail` can be used for `srcset`.

`woocommerce_gallery_thumbnail` has a width of 100px by default and `woocommerce_thumbnail` of 300px. That means that if a product image has a different aspect ratio than the default one defined in the Customizer, the 300px wide image is the biggest one that can be used! This is how a full-width Product Collection block looks:

![image](https://github.com/woocommerce/woocommerce/assets/3616980/7ca46c1a-df19-4023-994b-9fa3f8182ddd)

_(Note: you might need to open the image in a new tab to see what I mean)_

The Beanie and Belt products have square images, so they benefit from thumbnails that keep the aspect ratio and they have "bigger alternatives" to display. However, the Blouse product image is not square, so only the `woocommerce_thumbnail` and `woocommerce_gallery_thumbnail` thumbnails are available, and that's why it looks blurry. :monocle_face: 

A different (but related) issue is that thumbnails on the Single Product page use the `woocommerce_gallery_thumbnail` image without providing any `srcset`, that makes them very blurry:

![image](https://github.com/woocommerce/woocommerce/assets/3616980/120c7f52-fa4c-492e-82dd-0dc2ef112247)

### Next steps

This was not an in-depth investigation, I only took a look at the Single product page and the Product Collection blocks, so there might be more issues in other blocks. However, I think that the next steps cover a wide range of the issues we currently have with product images in WooCommerce.

So without further ado, these are the next steps:

#### Engineering

1. Thumbnail images in the single product template should have a `srcset` attribute that lets the browser choose the best image to load. (Issue created: https://github.com/woocommerce/woocommerce/issues/48431)
2. I think we should investigate adding new thumbnail sizes for product images. Given that several blocks rely on "cropped" images by default, having 100px and 300px thumbnails is not enough. We should evaluate which other sizes can be introduced and what are the pros and cons. I assume introducing new thumbnail sizes comes with the cost of requiring more disk space.
3. Evaluate if the `woocommerce_thumbnail` and `woocommerce_gallery_thumbnail` sizes are the appropiate or not. With many themes displaying the gallery thumbnails bigger than 100x100px, and with many devices having high-definition screens, I think `woocommerce_gallery_thumbnail` default to 100x100px is too low. We should investigate what's a good new value and change it in a backwards-compatible way. Same for `woocommerce_thumbnail`.
4. Find a way so we don't need to load the full-size product image when loading the Single Product page (see some ideas at the beginning of this comment).

#### Product/designers

1. It would be good to start investigating how we can display the WooCommerce Product Images settings from the Customizer in block themes.

**Comment by dinhtungdu** [comment]
#### About the full-size loading issue of thumbnails

I checked the original report but couldn’t reproduce the issue. I even tried the combination at the creation time of this issue:

- WordPress 5.6
- WooCommerce 8.0.0
- Storefront 3.5.0

But the thumbnail images on the catalog page are loading as expected, both on desktop and mobile. If anyone still has this issue, please share the reproducible steps with me!


#### About the full-size image loading by `jquery.zoom.js` on the single product page

I don’t think it’s an issue. As @Aljullu  mentioned above, that image doesn’t block the initial page render. I also did some tests with Lighthouse and didn’t see the full-size image in the report.


I can’t think of any other ideas than the ones listed above by Albert, but I'm not convinced to try one of them.

> * Downloading the full-size image only when the user interacts with the zoom. This has the drawback that it will probably add a delay until the zoomed-in image can be shown, as it won't have been downloaded before.

The drawback makes this idea not worth trying to me. As the content flashing issue is more serious than loading the full-size image after the DOM has been loaded.

> * Instead of using the full-size image for the zoom effect, use a smaller size (but still bigger than the thumbnail). This has the drawback that it would change how things work for existing stores, with the risk of breaking things. But I think it might be good to experiment with it if we can make it so it doesn't break backwards-compatibility.

With the big image size threshold [introduced](https://make.wordpress.org/core/2019/10/09/introducing-handling-of-big-images-in-wordpress-5-3/) in 5.3, this idea is automatically applied.

> * We could add `loading="lazy"` to the image. I don't think it would be a big improvement, as in most stores the product image is above-the-fold, so it would render on initial load anyway. But it might be worth trying.

As @Aljullu already mentioned, the image is above the fold most of the time, so the benefit is very subtle to me. Btw, to do so, we need to modify the way we initialize the zoom or even the library itself, which is likely to introduce backward compatibility issues.

So, my recommendation here is to leave the full-size image for zooming on the single product page as it is now.

**Comment by sunyatasattva** [comment]
I just want to add that even if the image is not render-blocking, loading a full-sized image on mobile before the user has shown any intent to zoom is just bad practice. As someone who travels often and needs to be conscious of data limits, I wouldn't feel happy about that.

**Comment by dinhtungdu** [comment]
> loading a full-sized image on mobile before the user has shown any intent to zoom is just bad practice. As someone who travels often and needs to be conscious of data limits, I wouldn't feel happy about that.

@sunyatasattva I agree with you. Thinking more about this issue, it's actually quite a challenging problem to solve:

On mobile device, shoppers can tap on the image to zoom in that area. So it's just one tap away to active the zoom feature. This make detecting shoppers' intent to zoom difficult.

I haven't had any idea to solve this problem yet. I can think of measuring the duration before shoppers scroll down the page and only load the full size image after a specific duration, but the duration is affected by so many factors like network speed and each shopper behavior. Any suggestion is more than welcome and greatly appreciated.

**Comment by Aljullu** [comment]
> 2. I think we should investigate adding new thumbnail sizes for product images. Given that several blocks rely on "cropped" images by default, having 100px and 300px thumbnails is not enough. We should evaluate which other sizes can be introduced and what are the pros and cons. I assume introducing new thumbnail sizes comes with the cost of requiring more disk space.
> 3. Evaluate if the `woocommerce_thumbnail` and `woocommerce_gallery_thumbnail` sizes are the appropiate or not. With many themes displaying the gallery thumbnails bigger than 100x100px, and with many devices having high-definition screens, I think `woocommerce_gallery_thumbnail` default to 100x100px is too low. We should investigate what's a good new value and change it in a backwards-compatible way. Same for `woocommerce_thumbnail`.

I dived deeper into this and changed my mind from my initial message. Instead of adding new cropped thumbnails, I think we should stop using cropped thumbnails altogether and switch to using WordPress-generated thumbnails (and crop them with CSS in the frontend). I expand on this in this GitHub discussion: https://github.com/woocommerce/woocommerce/discussions/49343.

**Comment by tjcafferkey** [comment]
Hmm, a few discussions happened off the back of this and we ended up deciding to leave it as is, since the large images are lazy loaded in. But here are a few of the reasons behind this decision:

1. They’re lazy loaded in, which would be recommended for performance reasons anyway.
2. All the thumbnails show, so a user can easily jump to ANY image within the gallery immediately. The large image is used as the main image so they all need to be ready to be activated into view.
3. The gallery doesn’t know how big the block its being rendered in is (e.g. it could be in 1/4 column or the entire width), only the viewport. This makes a scalable srcset/sizes solution work across mobile and desktop difficult.
4. The large images are also used for the zoom feature, so they’re ready when the user initiates zoom for a seamless experience instead of waiting for it to download.

The downside to this obviously is that we are sending more data over the wire, but having looked at various ecommerce stores online most (if not all) appear to be taking the same approach.

In short, if we don’t download the large images lazily we could compromise user experience by making them wait for them to download on demand.

I am going to close this issue for now but please feel free to reopen this again if you feel it requires further discussion or attention.
